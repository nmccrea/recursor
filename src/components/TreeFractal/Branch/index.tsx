import React from "react"
import { useSelector } from "react-redux"
import { selectAll } from "../../../state/branchPatterns/selectors"
import { BranchPattern, Depth } from "../../../state/branchPatterns/types"

/**
 * Generates a CSS object with styles dictated by the given branch pattern.
 */
const styleFor = ({ translation, scale, angle, color }: BranchPattern) => ({
  backgroundColor: color,
  transformOrigin: "center bottom",
  transform: [
    `translateY(-${100 * translation}%)`,
    `scale(${scale})`,
    `rotate(${angle}rad)`,
  ].join(" "),
})

interface BranchProps {
  branchPattern: BranchPattern
  currentDepth: Depth
}

const Branch = ({ branchPattern, currentDepth }: BranchProps) => {
  if (currentDepth >= branchPattern.depth) return null
  const branchPatterns = useSelector(selectAll)
  return (
    <div
      className={`tree ${branchPattern.id} depth-${currentDepth + 1}`}
      style={{
        position: "absolute",
        inset: 0,
        ...styleFor(branchPattern),
      }}
    >
      {branchPatterns.map((branchPattern) => (
        <Branch
          branchPattern={branchPattern}
          currentDepth={currentDepth + 1}
          key={`${branchPattern.id}`}
        />
      ))}
    </div>
  )
}

export default Branch
