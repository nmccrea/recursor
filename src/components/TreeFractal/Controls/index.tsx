import React from "react"
import { useSelector } from "react-redux"
import { selectIds } from "state/similarities/selectors"
import {
  Panel,
  GlobalControlsArea,
  SimilarityControlsArea,
  Instructions,
} from "./styled"
import GithubLink from "./GithubLink"
import CreateSimilarityButton from "./CreateSimilarityButton"
import SimilarityControls from "./SimilarityControls"

const Controls = () => {
  const similarityIds = useSelector(selectIds)
  return (
    <Panel>
      <GlobalControlsArea>
        <CreateSimilarityButton />
        <Instructions>{"< Add up to 3 similarities"}</Instructions>
        <GithubLink />
      </GlobalControlsArea>

      <SimilarityControlsArea>
        {similarityIds.map((similarityId) => (
          <SimilarityControls similarityId={similarityId} key={similarityId} />
        ))}
      </SimilarityControlsArea>
    </Panel>
  )
}

export default Controls
