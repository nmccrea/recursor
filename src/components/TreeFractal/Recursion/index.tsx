import React from "react"
import { useSelector } from "react-redux"
import { selectAll } from "../../../state/similarities/selectors"
import { Similarity, Depth } from "../../../state/similarities/types"

/**
 * Generates a CSS object with styles dictated by the given similarity.
 */
const styleFor = ({ translation, scale, angle, color }: Similarity) => ({
  backgroundColor: color,
  transformOrigin: "center bottom",
  transform: [
    `translateY(-${100 * translation}%)`,
    `scale(${scale})`,
    `rotate(${angle}rad)`,
  ].join(" "),
})

interface RecursionProps {
  similarity: Similarity
  currentDepth: Depth
}

const Recursion = ({ similarity, currentDepth }: RecursionProps) => {
  if (currentDepth >= similarity.depth) return null
  const similarities = useSelector(selectAll)
  return (
    <div
      className={`tree ${similarity.id} depth-${currentDepth + 1}`}
      style={{
        position: "absolute",
        inset: 0,
        ...styleFor(similarity),
      }}
    >
      {similarities.map((similarity) => (
        <Recursion
          similarity={similarity}
          currentDepth={currentDepth + 1}
          key={`${similarity.id}`}
        />
      ))}
    </div>
  )
}

export default Recursion
