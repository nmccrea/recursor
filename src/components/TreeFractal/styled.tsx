import styled from "styled-components"

const ControlsContainer = styled.div`
  position: fixed;
  top: 0px;
  left: 0px;
  width: 100vw;
  height: 100vh;

  display: flex;
  flex-direction: column;
  justify-content: flex-end;
  align-items: stretch;

  box-sizing: border-box;
  padding: 1rem;
`

export { ControlsContainer }
