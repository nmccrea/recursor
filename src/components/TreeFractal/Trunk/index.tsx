import React from "react"
import Branch from "../Branch"

const Trunk = () => (
  <div style={{ width: "10px", height: "200px", position: "relative" }}>
    <Branch
      currentDepth={0}
      similarity={{
        id: "trunk",
        color: "black",
        translation: 1,
        scale: 1,
        angle: 0,
        depth: Infinity,
      }}
    />
  </div>
)

export default Trunk
