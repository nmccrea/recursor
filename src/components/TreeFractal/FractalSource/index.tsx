import React from "react"
import Recursion from "../Recursion"

const FractalSource = () => (
  <div style={{ width: "10px", height: "200px", position: "relative" }}>
    <Recursion
      currentDepth={0}
      similarity={{
        id: "source",
        color: "black",
        translation: 1,
        scale: 1,
        angle: 0,
        depth: Infinity,
      }}
    />
  </div>
)

export default FractalSource
