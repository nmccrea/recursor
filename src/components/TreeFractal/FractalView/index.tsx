import React from "react"
import FractalSource from "./FractalSource"

const FractalView = () => (
  <div
    style={{
      width: "95vw",
      height: "70vh",
      display: "flex",
      justifyContent: "center",
      alignItems: "flex-end",
    }}
  >
    <FractalSource />
  </div>
)

export default FractalView
