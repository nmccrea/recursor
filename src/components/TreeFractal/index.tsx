import React from "react"
import { ControlsContainer } from "./styled"
import FractalView from "./FractalView"
import Controls from "./Controls"

const TreeFractal = () => {
  return (
    <>
      <FractalView />

      <ControlsContainer>
        <Controls />
      </ControlsContainer>
    </>
  )
}

export default TreeFractal
