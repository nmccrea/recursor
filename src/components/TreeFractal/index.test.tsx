import React from "react"
import { render } from "@testing-library/react"
import { Similarity } from "../../state/similarities/types"
import TreeFractal from "."

const MOCK_SIMILARITIES: { [id: string]: Similarity } = {
  "test/a": {
    id: "test/a",
    translation: 0.994,
    scale: 0.543,
    angle: -3.012 * Math.PI,
    depth: 3,
    color: "green",
  },
  "test/b": {
    id: "test/b",
    translation: 1.042,
    scale: 0.452,
    angle: 1.902 * Math.PI,
    depth: 2,
    color: "red",
  },
}

jest.mock("../../state/similarities/selectors", () => ({
  selectAll: () => Object.values(MOCK_SIMILARITIES),
}))

jest.mock("react-redux", () => ({
  useSelector: (selector: Function) => selector(),
}))

describe("<TreeFractal />", () => {
  it("renders correctly", () => {
    const { container } = render(<TreeFractal />)

    expect(container).toMatchSnapshot()
  })
})
