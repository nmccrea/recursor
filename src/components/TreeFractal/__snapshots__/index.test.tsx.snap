// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`\`<TreeFractal />\` renders correctly 1`] = `
.c3 {
  position: fixed;
  top: 0px;
  left: 0px;
  width: 100vw;
  height: 100vh;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-box-pack: end;
  -webkit-justify-content: flex-end;
  -ms-flex-pack: end;
  justify-content: flex-end;
  -webkit-align-items: stretch;
  -webkit-box-align: stretch;
  -ms-flex-align: stretch;
  align-items: stretch;
  box-sizing: border-box;
  padding: 1rem;
}

.c0 {
  position: fixed;
  top: 0px;
  left: 0px;
  width: 100vw;
  height: 100vh;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  overflow: hidden;
}

.c2 {
  -webkit-transform-origin: center bottom;
  -ms-transform-origin: center bottom;
  transform-origin: center bottom;
  position: absolute;
  top: 0px;
  bottom: 0px;
  left: 0px;
  right: 0px;
}

.c1 {
  width: 1vh;
  height: 15vh;
  position: relative;
}

.c4 {
  padding: 1rem;
  background-color: rgba(255,255,255,0.7);
  border-radius: 0.5rem;
  box-shadow: 0px 3px 1px -2px rgba(0,0,0,0.2), 0px 2px 2px 0px rgba(0,0,0,0.14),0px 1px 5px 0px rgba(0,0,0,0.12);
  height: 30%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: stretch;
  -webkit-box-align: stretch;
  -ms-flex-align: stretch;
  align-items: stretch;
  overflow-y: auto;
}

.c5 {
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  padding-bottom: 0.5rem;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c9 {
  display: grid;
  grid-template-columns: repeat(1,minmax(0,1fr));
}

.c7 {
  -webkit-flex: 1 0 0;
  -ms-flex: 1 0 0;
  flex: 1 0 0;
  margin-left: 1em;
  margin-right: auto;
}

.c8 {
  background-image: url(-- MOCK FILE EXPORT VALUE --);
  background-size: 1.6rem 1.6rem;
  width: 1.6rem;
  height: 1.6rem;
  opacity: 0.6;
  margin-right: 0.5rem;
}

.c8:hover {
  opacity: 0.8;
}

.c6 {
  border-radius: 0.5rem;
  box-shadow: 0px 3px 1px -2px rgba(0,0,0,0.2), 0px 2px 2px 0px rgba(0,0,0,0.14),0px 1px 5px 0px rgba(0,0,0,0.12);
  padding: 6px 16px;
  background-color: lightgray;
  color: white;
  -webkit-letter-spacing: 0.03em;
  -moz-letter-spacing: 0.03em;
  -ms-letter-spacing: 0.03em;
  letter-spacing: 0.03em;
  line-height: 1.75;
  text-transform: uppercase;
  background-color: cornflowerblue;
}

.c6:hover {
  box-shadow: 0px 2px 4px -1px rgba(0,0,0,0.2), 0px 4px 5px 0px rgba(0,0,0,0.14),0px 1px 10px 0px rgba(0,0,0,0.12);
  -webkit-filter: brightness(95%);
  filter: brightness(95%);
}

.c6:active {
  box-shadow: 0px 5px 5px -3px rgba(0,0,0,0.2), 0px 8px 10px 1px rgba(0,0,0,0.14),0px 3px 14px 2px rgba(0,0,0,0.12);
  -webkit-filter: brightness(100%);
  filter: brightness(100%);
}

@media (min-width:640px) {
  .c9 {
    grid-template-columns: repeat(2,minmax(0,1fr));
  }
}

@media (min-width:768px) {
  .c9 {
    grid-template-columns: repeat(3,minmax(0,1fr));
  }
}

@media (min-width:1024px) {
  .c9 {
    display: -webkit-box;
    display: -webkit-flex;
    display: -ms-flexbox;
    display: flex;
    -webkit-flex-direction: column;
    -ms-flex-direction: column;
    flex-direction: column;
  }
}

<div>
  <div
    class="c0"
  >
    <div
      class="c1"
    >
      <div
        class="c2 similarity-root depth-0"
        style="background-color: black; transform: translateY(-0%) scale(1) rotate(0rad);"
      />
    </div>
  </div>
  <div
    class="c3"
  >
    <div
      class="sc-AxhUy c4"
    >
      <div
        class="c5"
      >
        <button
          class="sc-fzoLsD c6"
        >
          Add
        </button>
        <span
          class="c7"
        >
          &lt; Add up to 3 similarities
        </span>
        <a
          href="https://github.com/mockiemcgee/test"
          target="blank"
        >
          <div
            class="c8"
          />
        </a>
      </div>
      <div
        class="c9"
      />
    </div>
  </div>
</div>
`;
